Extend Node-API to `libnode`

mmomtchev opened the pull request:
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

This PR adds support for instantiation and further interaction with an embedded version of Node.js (libnode) entirely through Node-API.

Refs: #43516

addaleax commented on doc/api/n-api.md:
```diff
@@ -515,6 +515,8 @@ the currently running Agent which was set by means of a previous call to
 `napi_set_instance_data()` will be overwritten. If a `finalize_cb` was provided
 by the previous call, it will not be called.
 
+Not compatible with `libnode`.
```
What would this mean?

mmomtchev replied:
That API call is not to be used when using `libnode`

addaleax replied:
Why not?

addaleax replied:
Yes, but that doesn’t mean that this call is not compatible with libnode or with embedded mode.

addaleax commented on doc/api/n-api.md:
```diff
@@ -6267,6 +6269,95 @@ idempotent.
 
 This API may only be called from the main thread.
 
+## Using Node.js as a shared library (`libnode`)
```
This has nothing to do with shared library mode.

mmomtchev replied:
This API allows using `libnode` through Node-API and it has no other uses.

addaleax replied:
That’s … wrong? Embedding and usage as a shared library are conceptually related but technically orthogonal.

mmomtchev replied:
Ok, let's call it `Using an embedded version of Node.js`

addaleax commented on doc/api/embedding.md:
```diff
@@ -160,8 +160,46 @@ int RunNodeInstance(MultiIsolatePlatform* platform,
+<!--introduced_in=REPLACEME-->
+
+As an alternative, an embedded Node.js can also be fully controlled through
+Node-API. This API supports both C and C++ through [node-addon-api][].
```
If we do this, we should probably start adding converters between some Node-API values and V8/Node.js C++ API values (`napi_value` ↔ `Local<Value>`, `napi_env` ↔ `Environment*`/`Isolate*`).

The reason nobody has implemented a Node-API embedder API yet is that as an embedder, you often need fine-grained control over a lot of V8 options, and it’s not very practical to mirror the entire V8/Node.js embedder API to C (and that the Node.js C++ embedder API itself still has some rough edges that should probably be ironed out first).

mmomtchev replied:
If you need to use raw V8 primitives then you can't use Node-API - this applies to both Node.js -> C++ (binary node addons) and to C++ -> Node.js (libnode). Still this API covers lots of use cases - mostly C++ software that needs to accept JS plugins. It does not cover Electron which has very particular needs.
Allowing for more fine-grained control over the initialisation is of course something that should be considered.

addaleax replied:
> If you need to use raw V8 primitives then you can't use Node-API

Right, that’s why I’m bringing this up.

mmomtchev replied:
Should you be able to mix V8 primitives and Node-API is another topic. If you take the ABI compatibility out of Node-API, there remains little use for it. However `node-addon-api` is a truly higher level API and it is much more practical to use from C++ then raw V8. Maybe there is a use case for software which has 95% `node-addon-api` and 5% custom V8 code. But this is something that goes beyond this PR - and it probably can be used both by embedders and addon authors.

NickCarducci replied:
I believe Cloudflare ([service](https://github.com/cloudflare/worker-emscripten-template)) Workers will need [converters](https://markethistory.quora.com/What-is-the-relationship-between-node-js-and-C-1). Is this a Private API thing?

addaleax commented on src/js_native_api_v8.cc:
```diff
@@ -775,6 +802,121 @@ napi_status NAPI_CDECL napi_get_last_error_info(
+  CHECK_ARG(env, env);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  {
+    int r = node::SpinEventLoop(node_env->node_env()).FromMaybe(1);
+    if (exit_code != nullptr) *exit_code = r;
+    node::Stop(node_env->node_env());
+  }
```
It’s going to be heavily counterintuitive that this is part of a destruction operation.

mmomtchev replied:
Once you have drained the event loop, you have no other course of action then to destroy the environment - this is the reason I did it this way. There is no point in having a separate API call, if you will always have to call both of them one after the other. The real question is can we (and should we) keep an environment when its event loops is empty? This would allow to call async functions in a persistent environment but it will introduce a significant difference between the stand-alone Node.js and a embedded Node.js.

addaleax replied:
You could want to take an action in response to the event loop being drained, though? 

> The real question is can we (and should we) keep an environment when its event loops is empty? 

I think the answer to both questions here is yes.

mmomtchev replied:
Ok, but this would require splitting `node::SpinEventLoop` into iteration and cleanup and it is something that is not possible with the current embedders' API.

addaleax replied:
… but it is? The docs for `SpinEventLoop` say which steps it takes, and all of those are public API. `SpinEventLoop()` is a helper, not a building block, and exists only for convenience and stronger API stability for those for whom it is sufficient.

mmomtchev replied:
Ok, I added `napi_run_environment` that spins up the event loop without destroying the environment when it is emptied and a unit test with an async callback.

addaleax reviewed (changes requested):
I would strongly encourage you to go through the current embedder API and understand how embedders use it. I would love to see a C embedding API, but since the goal of Node-API-style APIs is to provide long-term stability, it’s going to need to be an API that is designed to account for multiple usage patterns (e.g.: integration with libuv, usage with a node-addon-api-style C++ wrapper) and probably some kind of API versioning in advance (since Node.js and V8 embedding changes their API definitions much more frequently than JS changes as a language).

mmomtchev commented:
@KevinEady
1.  I agree that this API would not cover all use cases - in particular Electron has very specific requirements - but if you look at these exchanges, you will see that even their team was interested in some form of ABI or at least ABI stability. Still, such an API will be sufficient for a very significant portion of the uses cases. What needs discussing is:
  * How can greater customisability of the engine startup be supported and where should the line be drawn
  * Should we support mixed use with conversions between `v8::Local` and `napi`
2. I don't think that `napi_platform` is V8-specific - every engine would have a common main initialisation procedure and will support a single or many independent environments.

KevinEady commented:
We discussed in the 1 July 2022 Node API meeting, and we did want to bring up again the comment we had before:
> It may be beneficial to create some APIs that are outside of the Node-API scope, but are still C-based and provide ABI stability in order to perform this embedding scenario, where we can ignore the platform-agnostic requirement, ie. inside a different header file.

There is a logical separation between "module extensions API" and "node hosting API" so it may be beneficial to separate this embedding API into a separate header.

Regarding:
> Should we support mixed use with conversions between v8::Local and napi

Can you clarify this a bit?

mmomtchev commented:
> > Should we support mixed use with conversions between v8::Local and napi
> 
> Can you clarify this a bit?

Public API for retrieving the V8 reference from a `napi*` handle and a constructor for creating a `napi*` handle from a V8 reference. It goes against the Node-API logic, but it would allow extending Node-API without modifying it.

mmomtchev commented:
@KevinEady @addaleax @rvagg  My goal is to be able to support installation of this (`libnode` + this API) along a NodeSource Node.js package

Basically, I have two choices:
* Either `libnode` & `libnode-dev` are completely independent of `nodejs` and install headers to `/usr/include/libnode`
* Either I install this single new header file `/usr/include/node` - and then I require that `nodejs` is present and it is a specific version

Whatever I do at this point will likely stay. Do you have any opinion?

addaleax commented:
@mmomtchev Two things:

- I would not split header files from the existing header bundle, since the default for embedders will still be using the regular C++ APIs that are also used by addons
- I would consider shipping the shared library variant of Node.js completely separate and independent from any embedding API additions or changes

mmomtchev commented:
@addaleax And do you have an idea for an elegant solution for having top-level functions appear in the global object the way they do in the REPL, besides `vm.runInThisContext()`? Ideally I would like to hide this from the end user.

addaleax commented:
@mmomtchev If somebody wants that, they can just do the same thing the REPL does to make that happen, right?

mmomtchev commented:
@addaleax I have a problem with restarting the event loop since `NodePlatform::DrainTasks` :

https://github.com/nodejs/node/blob/3738b57102b5fad003e9e3ec7c936136980556bc/src/node_platform.cc#L440

can be run only once per isolate. If I don't drain the tasks, some pending Promises (the dynamic `import` for example) do not get resolved. If I drain them, they do not get rescheduled. This is valid even when using the existing embedders' API. Do you have any opinion about it?

mhdawson commented:
I was experimenting with using Node.js to support WASM containers. For that I needed to use libnode with [crun](https://github.com/containers/crun).  Unfortunately crun is current a C project and integrating the C++ API from libnode was going to be difficult (if crun would even consider moving to C++ versus C). 

The model is to dlsym the required methods as crun support multiple engines and only wants to load the shared libraries needed.  

With this PR is was relatively straight forward to get it working by using the new methods to start/get an env and then existing node-api methods to create and start a script to run.

At this for the simple `run a script` case it seemed to be enough.  was experimenting with using Node.js to support WASM containers. For that I needed to use libnode with [crun](https://github.com/containers/crun).  Unfortunately crun is current a C project and integrating the C++ API from libnode was going to be difficult (if crun would even consider moving to C++ versus C). 

The model is to dlsym the required methods as crun support multiple engines and only wants to load the shared libraries needed.  

With this PR is was relatively straight forward to get it working by using the new methods to start/get an env and then existing node-api methods to create and start a script to run.

At this for the simple `run a script` case it seemed to be enough. 

You can see what it used here:  https://github.com/mhdawson/crun/blob/node-wasm-experiment/src/libcrun/handlers/wasmtime.c

vmoroz commented on src/js_native_api.h:
```diff
@@ -102,6 +102,34 @@ node_api_symbol_for(napi_env env,
 #endif  // NAPI_EXPERIMENTAL
+
+#ifdef NAPI_EMBEDDING
+NAPI_EXTERN napi_status NAPI_CDECL napi_create_platform(int argc,
```
Currently the Node-API is split up between two files: `js_native_api.h` and `node_api.h`.
The difference is that the `js_native_api.h` contains only cross-engine JS-specific APIs and `node_api.h` has Node.JS-specific APIs. 
We use `js_native_api.h` in different scenarios where we want to interact with different JS engine using ABI-safe Node-API functions. For example, see the [Babylon Native](https://github.com/BabylonJS/BabylonNative/tree/master/Dependencies/napi), [Hermes Node-API](https://github.com/microsoft/hermes-windows/blob/main/API/hermes/hermes_napi.cpp) or [V8-JSI](https://github.com/microsoft/v8-jsi/tree/master/src/napi). In all these scenarios we implement `js_native_api.h` APIs, and we do not use Node.JS or Node-specific `node_api.h`. Even the `js_native_api_v8.h` and `js_native_api_v8.cc` must be free from any Node.JS specific code. Any interaction with the Node-JS specific code must be done either through the functions declared in the `js_native_api_v8_internals.h` or overriding virtual `napi_env__` in the derived `node_napi_env__` class in the `node_api_internals.h`.
The `node_api.h` has the Node.js specific functions and mostly contains code concerning Node-JS modules and asynchronous code.
From this perspective, the Node.JS embedding API must not be part of the `js_native_api.h` header file. It could be in the `node_api.h`, but since this file is to be used by modules, the best option is to put these new APIs in their own `node_hosting_api.h` or `node_embedding_api.h` file. Other applications, such as [Deno](https://deno.land/), could implement the full support for `node_api.h` and `js_native_api.h` to reuse the Node-API modules, but they would never implement the same Node.JS embedding APIs.
These embedding APIs are Node.JS specific and cannot be re-used for other engines or applications. E.g. other engines do not create a thread pool on their own, or cannot handle Node.JS CLI arguments,
It could be either pa

vmoroz commented on src/js_native_api.h:
```diff
@@ -102,6 +102,34 @@ node_api_symbol_for(napi_env env,
+napi_destroy_platform(napi_platform platform);
+
+NAPI_EXTERN napi_status NAPI_CDECL
+napi_create_environment(napi_platform platform,
```
Please note that the term `Environment` is somehow overused in Node.JS. 
The `napi_env` is different from the Node.JS `Environment`. 
We create one instance of `napi_env` per each native module. It allows using the `napi_env` instance settings and other flags per module. Each Node.JS environment may have zero or more `napi_env` instances depending on how many Node-API modules it loads.
It is important to clarify what type of environment the `napi_create_environment` function creates. I bet it is not the `napi_env`, but rather a moniker to the Node.JS `Envrironment`, which I guess must have a different type such as `napi_node_instance`, `napi_node_env`, or something else along those lines.

mmomtchev replied:
`napi_create_environment` creates a `napi_env` instance linked to a Node `Environment` containing an additional context pointer (`set_embedded`/`get_embedded`). In theory (one must take care to register the cleanup only once) there can also be multiple `napi_env` instances for a single embedded `Environment` - but I don't see any reason why.

The whole point of the NAPI embedding is that it allows reusing of the NAPI functions for the embedded environment. So whatever type I return, that type must remain compatible with all the NAPI functions. In the C++ API, I return a different type (inherited from the basic `Node::Env`), but I don't see any practical way to do this in C. If I return a different type, you will have to cast it every time you invoke a NAPI function.

I don't understand the problem with the current situation - every `napi_env` is linked to a single `Environment`. When embedding, it is the `Environment` that is slightly different, not the `napi_env`.

mmomtchev replied:
If I name this function `napi_create_instance` but still return a `napi_env`, does this address your concern?

mmomtchev replied:
Or there is the quite cumbersome option:
```c
napi_embedded_instance inst;
napi_create_instance(platform, NULL, NULL, &inst);
napi_value global;
napi_get_global(napi_instance_env(inst), &global);
```

It is ugly in C, but it will be covered up in C++

vmoroz commented on src/js_native_api_v8.cc:
```diff
@@ -773,6 +815,177 @@ napi_status NAPI_CDECL napi_get_last_error_info(
+                                            char** argv,
+                                            int exec_argc,
+                                            char** exec_argv,
+                                            char*** errors,
```
With the ABI-safe APIs we cannot allocate memory in one module and free it in another. E.g. in our Microsoft Office apps we use custom memory allocations with additional safety checks. The memory allocated in embedded Node.JS cannot be deallocated in Office code that may embed it. To be ABI-safe all memory must be allocated and freed in the same module.
For example, in the Node-API we often use the Finalizer callback that supports such single memory ownership pattern. In COM we use AddRef/Release to enforce this pattern too. 
In this case we must at least return a pointer to the finalizer function that must be called by `napi_create_platform` user after they finish processing errors, or have a special `napi_free_string_array`.
My personal preference is to have a special `write_error` callback that we call to write the errors to the hosting app. In that case we would not need to allocate memory and just give them `data()` and `length()` from our stored `std::string`. It will be up to the hosting app how to handle it - it can either print or store them. The `write_error` callback could be called multiple times - once per each error string. There is no need to "terminate" the list - the end of the `napi_create_platform` is a natural `termination`. The `write_error` callback must accept the `void* hostingData` which must be passed to the `napi_create_platform` to let the hosting application to encapsulate its context for the `write_error` callback.

mmomtchev commented on src/js_native_api_v8.cc:
```diff
@@ -773,6 +815,177 @@ napi_status NAPI_CDECL napi_get_last_error_info(
   return napi_ok;
 }
 
+#define HANDLE_ERRORS_VECTOR(errors, vec)                                      \
```
Done, even if `errors` is mostly wishful thinking at the moment - in reality Node will simply abort on failure - there is no unit testing for this reason

mmomtchev committed:
node-api: libnode embedding

mmomtchev committed:
new linter rules

mmomtchev committed:
remove unintented change

mmomtchev committed:
restore require/import in main_script

jasongin commented on doc/api/n-api.md:
````diff
@@ -6310,6 +6310,147 @@ idempotent.
+> Stability: 1 - Experimental
+
+```c
+napi_status napi_destroy_platform(napi_platform platform, int *exit_code);
````
`napi_destroy_environment` has an `exit_code` parameter; `napi_destroy_platform` does not.

jasongin replied:
Don't forget to fix the doc for `napi_destroy_environment` also.

mmomtchev committed:
rectify napi_destroy_platform

jasongin commented on src/js_native_api_v8.cc:
```diff
@@ -692,6 +734,181 @@ napi_status NAPI_CDECL napi_get_last_error_info(
+    return napi_closing;
+
+  *result =
+      v8impl::JsValueFromV8LocalValue(scope.Escape(promise_object->Result()));
```
`SpinEventLoopWithoutCleanup` can return for either of two reasons:
  1. The condition lambda returns false (because the promise is resolved)
  2. There are no pending callbacks.

In the second case, the call to `promise_object->Result()` here may terminate the process with `SIGABRT` and the following error message:
```
FATAL ERROR: v8_Promise_Result Promise is still pending
```

I encountered that error from passing in a promise with an empty resolve callback -- my intention was to just have it run indefinitely for now, and I'd consider graceful cleanup later.

Should this code check the promise state again, and return a different error code (instead of aborting the process) if the promise is not resolved?

mmomtchev replied:
Node.js in general has an absolutely horrible behavior when it comes to `Promise`s that cannot be resolved because they have no handlers - it always simply quits - without an error code. This has already been discussed and it seems that it will stay like this. Anyway, the other possible valid behavior is to simply hang forever. So, in this case, besides suppressing the error message and the error code, there isn't much that can be done.

mmomtchev committed:
catch up with main

mmomtchev committed:
the linter was not happy with this indentation

mhdawson commented:
> I would strongly encourage you to go through the current embedder API and understand how embedders use it. I would love to see a C embedding API, but since the goal of Node-API-style APIs is to provide long-term stability, it’s going to need to be an API that is designed to account for multiple usage patterns (e.g.: integration with libuv, usage with a node-addon-api-style C++ wrapper) and probably some kind of API versioning in advance (since Node.js and V8 embedding changes their API definitions much more frequently than JS changes as a language).

I think having either an Experimental version of the C embedding API as part of Node-API, or possibly a Experimental version of the C embedding API which uses Node-API but which is not part of it to start would be a good way to get feedback so that we can work through the issues mentioned.  @addaleax would having the C API outside of the Node-API to start help to address some of your concerns about the potential faster pace of change?  In that case the path would be
1. C-APIs, using Node-API for types etc, but separate headers and naming to make it clear the APIs are not covered by the ABI stability guarrantee to start. 
1. Try to raise awareness of the APIs (through social, etc.)  to get as many people to try them out, provide feedback. 
1. At some later point we could move those APIs into Node-API once we are comfortable we have enough experience to believe they are unlikely to need to change significantly/often.

This would let us see the rate/pace of change of the C APIs over time through GitHub history as part of figuring out if/when we are ready to move to ABI stable. The same would be true if we introduced them as experimental but the initial separation might make it more obvious/clearer.

mmomtchev commented:
Alas, lately, the main theme of all my social media accounts is the fact that I barely have enough money to eat and I have been served an eviction notice - because I have been unemployed since 2020. This happens after my previous employers tried to convince me that I was suffering from schizophrenia because I was hallucinating that people around me were posting dicks after a false rape accusation that was passed through the French criminal justice system. After I tried suing my employers, my lawyer kept telling me that he was waiting for the court to schedule a hearing, but in fact he had rendered all proceedings void. Since this day, I am having psychotic episodes at each interview - I usually have the feeling that the companies that try to hire me post the same images as my previous employers.

Given this rather particular situation of mine, I am afraid that my social media accounts wouldn't be the ideal location for this new API to gain any traction. However the original issue has a number of thumbs up, and it would probably be a much better place.

vmoroz commented:
@mmomtchev, may I help you to complete this PR?
We had discussed this PR in the Node-API meetings and we believe that this work can be quite beneficial in a number of Node.js embedding scenarios.
I can submit changes directly to your PR (the preferred way) or open a new PR (less preferred option).
Let me know which way you would prefer to go.
In any case, you are the author of this PR, and my part will be just adding a few finishing touches to get it merged.

mmomtchev committed:
the underlying libuv issue has been resolved

mmomtchev committed:
remove the deps/uv patch

mmomtchev commented:
@vmoroz You are welcome to push any modifications

I catched up with `main` and removed the `deps/uv` patch now that libuv has been updated.

There are a couple of issues that needs discussing such as the path. The embedded environment is neither an ES6 nor a CJS environment and does not really have a main script. I use Node.js' internal cross-platform exec path detection so that all `require` and `import` are relative to the binary of the main program that is linked to `libnode`. This variable gets passed as an argument in the bootstrap where it becomes a global symbol and I see that it clashes with `mksnapshot`. Both decisions - looking for modules relative to the binary - and passing this path as a new global symbol - need to be discussed.

I have also added the option for specifying an API level from the C/C++ API. The dynamic API level and the new realms are the two significant new changes that have the most repercussions for this PR.

mhdawson commented:
From discussion in the Node-api team meeting today, without looking to closely at the details it seemed reasonable that the APIs should take the path for which require and import are relative to as a parameter.

If the embedder wants to make that the binary Node.js is embedded into they can, but it also provides the option to do something else if that makes sense.

In terms of a news symbol we also thought that pretending that there was a main script by setting existing values such that it looks like a main script was run from the path provided to the API by the embedder and a pre-defined name for the main script might be an option. Does that make any sense?

mmomtchev committed:
remove the custom main script arg and use execPath

mmomtchev commented:
@mhdawson I have removed the custom `path` argument to the bootstrapper script that was causing a symbol collision. Normally, the embedded environment uses as much as possible from the normal `import` and `require` chains which means that it can be configured by using the `NODE_PATH` variable and should follow the usual directory relative to the `package.json` convention.

mmomtchev commented:
@joyeecheung I can add another `exec_path` to `Environment` - lets call it `modules_root` that will implement configuring the root from NAPI and carry it in the `process` object. I can also simply modify the `exec_path` in this particular case, but it is not a very clean solution as it reuses an existing parameter with a new meaning.

I cannot derive from `Environment` without making a huge mess as it has its own factory. I will have to implement it in `Environment`, but use it only for when there is an embedded environment.

From what I understand, the decision to not support `NODE_PATH` for ES6 modules was a deliberate one. There is no official way of setting the root of ES6 modules?

mmomtchev committed:
Update lib/internal/bootstrap/switches/is_embedded_env.js

Co-authored-by: Gabriel Schulhof <gabrielschulhof@gmail.com>

mmomtchev commented on Makefile:
```diff
@@ -287,6 +287,8 @@ coverage-report-js:
 cctest: all
 	@out/$(BUILDTYPE)/$@ --gtest_filter=$(GTEST_FILTER)
 	@out/$(BUILDTYPE)/embedtest "require('./test/embedding/test-embedding.js')"
+	@out/$(BUILDTYPE)/napi_embedding "require('./test/embedding/test-napi-embedding.js')"
```
@gabrielschulhof this is the name of the unit test, most unit tests / benchmarks that test Node-API use the `napi` abbreviation

mmomtchev committed:
adapt to the new embedding API

mmomtchev commented:
I adapted this to the latest changes in the embedding initialization which impacted my API as well - `napi_create_platform` and `napi_create_environment` . I opened #49512 and #49513 for the two unit test failures which are not related to this PR.

samardzicneo commented:
May one ask what the holdup is on this PR? Has OP lost interest? I'm part of a project that relies on this API and we're not too keen on using an unofficial fork and having to build Nodejs manually for all the platforms we support.

mhdawson commented:
@samardzicneo do you have cycles to help get it across the line?  I think @vmoroz also showed interest in moving this forward. @mmomtchev are you still going to have cycles or would it be good to have either @samardzicneo, @vmoroz or both try to move it forward?

samardzicneo commented:
> @samardzicneo do you have cycles to help get it across the line?  I think @vmoroz also showed interest in moving this forward. @mmomtchev are you still going to have cycles or would it be good to have either @samardzicneo, @vmoroz or both try to move it forward?

Really, I'd love to do so, especially since I need this for a project of mine, but honestly, I don't trust my C/C++ skills enough to be sending PRs to Node 😅

Believe me, you don't want my messy code anywhere near your runtime.

vmoroz committed:
move code to node_api_embedding.cc

vmoroz commented:
@mmomtchev , I have moved the embedded code out of js_native_api_v8.cc to the new file node_api_embedding.cc. The main reason is that we support a scenario where the js_native_api_v8.cc can be used outside of Node.js code on its own.
(In node_api_embedding.cc line 81 is changed to have an explicit static_cast to suppress MSVC warning for int64_t to int32_t conversion).
Please review the change.

jasongin commented on src/node_api_embedding.cc:
```diff
@@ -0,0 +1,254 @@
+  // and the v8::locker
+  delete emb_env;
+
+  cppgc::ShutdownProcess();
```
This call to `ShutdownProcess()` prevents another environment instance from being created in the same process. Specifically, a second call to `napi_create_environment()` crashes the process:
```
# Check failed: (g_page_allocator) != nullptr.
``` 
I found this because the [node-api-dotnet project](https://github.com/microsoft/node-api-dotnet/pull/219) runs multiple embedding test cases in a process, and each test case uses a separate environment instance. With this line removed, the latest PR revision works well.

Was there a reason to add this call to `ShutdownProcess()`?

vmoroz committed:
comment typo fix

vmoroz commented:
@mmomtchev , the code is rebased on the latest main code.

vmoroz committed:
delete compiled binary file from test

vmoroz committed:
Remove ShutdownProcess from napi_destroy_environment

Implemented by @jasongin

vmoroz commented:
> Any updates?

@reitowo , please check the PR #54660 for the progress and the related discussion #23265.
I wonder what is your scenario for the new API and what type of embedding customizations you are looking for?

reitowo commented:
I was coming from the doc of microsoft's project node & .net interop, just to check the status of this. ☺️
